pkg/recommendation/recommender/resource/registry.go

```diff
@@ -36,118 +54,137 @@ func (rr *ResourceRecommender) Name() string {
+		oomHistoryLengthDuration, err := time.ParseDuration(oomHistoryLength)
+		if err != nil {
+			return err
+		}
+		rr.OOMHistoryLength = oomHistoryLengthDuration
 	}
 
-	oomProtection, exists := recommender.Config["oom-protection"]
-	if !exists {
-		oomProtection = "true"
+	if oomBumpRatio, exists := config["oom-bump-ratio"]; exists {
+		oomBumpRatioFloat, err := strconv.ParseFloat(oomBumpRatio, 64)
+		if err != nil {
+			return err
+		}
+		rr.OOMBumpRatio = oomBumpRatioFloat
 	}
+	return nil
+}
 
-	oomProtectionBool, err := strconv.ParseBool(oomProtection)
+// NewResourceRecommender create a new resource recommender.
+func NewResourceRecommender(recommender apis.Recommender, recommendationRule *analysisv1alph1.RecommendationRule, oomRecorder oom.Recorder) (*ResourceRecommender, error) {
+	if recommender.Config == nil {
+		recommender.Config = map[string]string{}
+	}
+	// The priority of parameter is: config in RecommendationRule > RecommendationConfiguration in configmap > default value
+	defaultSpecificationBool, err := strconv.ParseBool(defaultSpecification)
 	if err != nil {
 		return nil, err
 	}
 
-	oomHistoryLength, exists := recommender.Config["oom-history-length"]
-	if !exists {
-		oomHistoryLength = "168h"
+	defaultSpecifications, err := GetResourceSpecifications(DefaultSpecs)
+	if err != nil {
+		return nil, err
 	}
 
-	oomHistoryLengthDuration, err := time.ParseDuration(oomHistoryLength)
+	defaultOOMProtectionBool, err := strconv.ParseBool(defaultOOMProtection)
 	if err != nil {
 		return nil, err
 	}
 
-	OOMBumpRatio, exists := recommender.Config["oom-bump-ratio"]
-	if !exists {
-		OOMBumpRatio = "1.2"
+	defaultOOMHistoryLengthDuration, err := time.ParseDuration(defaultOOMHistoryLength)
+	if err != nil {
+		return nil, err
 	}
 
-	OOMBumpRatioFloat, err := strconv.ParseFloat(OOMBumpRatio, 64)
+	OOMBumpRatioFloat, err := strconv.ParseFloat(defaultOOMBumpRatio, 64)
 	if err != nil {
 		return nil, err
 	}
 
-	return &ResourceRecommender{
+	rr := &ResourceRecommender{
 		*base.NewBaseRecommender(recommender),
-		cpuSampleInterval,
-		cpuPercentile,
-		cpuMarginFraction,
-		cpuTargetUtilization,
-		cpuHistoryLength,
-		memSampleInterval,
-		memPercentile,
-		memMarginFraction,
-		memTargetUtilization,
-		memHistoryLength,
+		defaultCpuSampleInterval,
+		defaultCpuPercentile,
+		defaultCpuMarginFraction,
+		defaultCpuTargetUtilization,
+		defaultCpuHistoryLength,
+		defaultMemSampleInterval,
+		defaultMemPercentile,
+		defaultMemMarginFraction,
+		defaultMemTargetUtilization,
+		defaultMemHistoryLength,
 		oomRecorder,
-		oomProtectionBool,
-		oomHistoryLengthDuration,
+		defaultOOMProtectionBool,
+		defaultOOMHistoryLengthDuration,
 		OOMBumpRatioFloat,
-		specificationBool,
-		specifications,
-	}, nil
+		defaultSpecificationBool,
+		defaultSpecifications,
+	}
+
+	// Override the default value with RecommendationConfiguration in configmap
+	if err = SetRecommendationConfig(rr, recommender.Config); err != nil {
```

Comment:
1. merge recommender.Config and recommendationRule.Spec.Config first
2. do other initialization steps

Reply:
1. if the value in recommender.Config and recommendationRule.Spec.Config conflict, I will use the value of recommendationRule. Is that right?

Reply:
> 1. if the value in recommender.Config and recommendationRule.Spec.Config conflict, I will use the value of recommendationRule. Is that right?

yes, recommendationRule.Spec.Config have the highest priority.

Reply:
also, api change PR should be merged first and use the newest api version in crane

Reply:
ok